### ISSUE TITLE
Keep array items in sync with Polymer templates

### ISSUE THREAD

**Issue by heruan** [vaadin/flow#3399]
I'm experiencing a use cases with Polymer templates where server/client models diverge: if the model has a list property, when an item property is changed on the client the server isn't notified.

```java
interface MyListModel extends TemplateModel {

    List<Item> getItems();

}
```

```js
class MyList extends Polymer.Element {

    static get properties() {
        return {
            items: Array
         };
      }

}
```

```html
<template>
    <dom-repeat items="[[items]]">
        <vaadin-text-field label="[[item.label]]" value="{{item.value}}"></vaadin-text-field>
    </dom-repeat>
</template>
```

Then, after editing some values, on the server:

```java
this.getModel().getItems().forEach(item -> getLogger().info(item.getValue()));
```

All the logger value are the ones before editing.

**Comment by heruan** [comment]
This web component is an example of binding which works in Polymer using `MutableData` mixin, but not in Flow:
```html
<link rel="import" href="../polymer/polymer-element.html">
<link rel="import" href="../polymer/lib/mixins/mutable-data.html">

<dom-module id="my-list">
  <template>
    <button on-click="notifyItems">Notify</button>
    <ul>
      <template is="dom-repeat" items="[[items]]" mutable-data>
        <li>Node #[[item.id]] ([[item.x]], [[item.y]])</li>
      </template>
    </ul>
  </template>

  <script>
    class MyList extends Polymer.MutableData(Polymer.Element) {

      static get is() {
        return 'my-list';
      }

      static get properties() {
        return {
          items: {
            type: Array,
            notify: true,
            value: [
              {
                id: 1,
                x: 10,
                y: 20
              }
            ]
          }
        };
      }

      notifyItems() {
        this.simulateExternalLibrary();
        this.notifyPath('items');
      }

      simulateExternalLibrary() {
        this.items[0].x = 30;
      }
    }
    window.customElements.define(MyList.is, MyList);
  </script>
</dom-module>
```

Having both `extends Polymer.MutableData(Polymer.Element)` and `<template ... mutable-data>` makes the bindings to update when specific item properties are modified by an external library and `notifyPath('items')` is invoked. With Flow, the server-side items have the old property values instead.

**Comment by heruan** [comment]
Potentially depends on #2413 (which might be the root cause for this).

**Comment by denis-anisimov** [comment]
This is duplicate of #2413.
And also it's a duplicate of #4091 (this is has a PR).
Please reopen if you think there are uncovered things in #2413.
Also you may check the IT test case in PR for #4091. It's almost exactly the same as yours except native `input` is used instead of `vaadin-text-field`.
